capabilities/writetarget/beholder/pb/data-feeds/on-chain/registry/decode.go

```diff
@@ -0,0 +1,195 @@
+	}
+
+	// Allocate space for the messages (event per updated feed)
+	msgs := make([]*FeedUpdated, 0, len(*reports))
+
+	// Iterate over the underlying Mercury reports
+	for _, rf := range *reports {
+		// Notice: we assume that Mercury will be the only source of reports used for Data Feeds,
+		// at least for the foreseeable future. If this assumption changes, we should check the
+		// the report type here (potentially encoded in the feed ID) and decode accordingly.
+
+		// Decode the common Mercury report
+		rm, err := mercury_vX.Decode(rf.Data)
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode Mercury report: %w", err)
+		}
+
+		// Parse the report type
+		t := mercury_vX.GetReportType(rm.FeedID)
+
+		// Notice: we publish the DataFeed FeedID, not the unrelying DataStream FeedID
+		feedID := datafeeds.FeedID(rf.FeedID)
+
+		switch t {
+		case uint16(3):
+			rm, err := mercury_v3.Decode(rf.Data)
+			if err != nil {
+				return nil, fmt.Errorf("failed to decode Mercury v%d report: %w", t, err)
+			}
+
+			msgs = append(msgs, &FeedUpdated{
+				// Event data
+				FeedId:                feedID.String(),
+				ObservationsTimestamp: rm.ObservationsTimestamp,
+				Benchmark:             rm.BenchmarkPrice.Bytes(), // Map big.Int as []byte
+				Report:                rf.Data,
+
+				// Notice: i192 will not fit if scaled number bigger than f64
+				BenchmarkVal: toBenchmarkVal(feedID, rm.BenchmarkPrice),
+
+				// Head data - when was the event produced on-chain
+				BlockHash:      m.BlockHash,
+				BlockHeight:    m.BlockHeight,
+				BlockTimestamp: m.BlockTimestamp,
+
+				// Transaction data - info about the tx that mained the event (optional)
+				// Notice: we skip SOME head/tx data here (unknown), as we map from 'platform.write-target.WriteConfirmed'
+				// and not from tx/event data (e.g., 'platform.write-target.WriteTxConfirmed')
+				TxSender:   m.Transmitter,
+				TxReceiver: m.Forwarder,
+
+				// Execution Context - Source
+				MetaSourceId: m.MetaSourceId,
+
+				// Execution Context - Chain
+				MetaChainFamilyName: m.MetaChainFamilyName,
+				MetaChainId:         m.MetaChainId,
+				MetaNetworkName:     m.MetaNetworkName,
+				MetaNetworkNameFull: m.MetaNetworkNameFull,
+
+				// Execution Context - Workflow (capabilities.RequestMetadata)
+				MetaWorkflowId:               m.MetaWorkflowId,
+				MetaWorkflowOwner:            m.MetaWorkflowOwner,
+				MetaWorkflowExecutionId:      m.MetaWorkflowExecutionId,
+				MetaWorkflowName:             m.MetaWorkflowName,
+				MetaWorkflowDonId:            m.MetaWorkflowDonId,
+				MetaWorkflowDonConfigVersion: m.MetaWorkflowDonConfigVersion,
+				MetaReferenceId:              m.MetaReferenceId,
+
+				// Execution Context - Capability
+				MetaCapabilityType:           m.MetaCapabilityType,
+				MetaCapabilityId:             m.MetaCapabilityId,
+				MetaCapabilityTimestampStart: m.MetaCapabilityTimestampStart,
+				MetaCapabilityTimestampEmit:  m.MetaCapabilityTimestampEmit,
+			})
```

Comment:
Maybe consider wrapping these `FeedUpdated` creations in constructors, especially with so much indentation.
```suggestion
			msgs = append(msgs, newFeedUpdatedV3(feedID, rm, rf, m))
```